doc/1chipML_manual.tex

```diff
@@ -307,7 +307,42 @@ \subsection{The gradient descent method}
 
 \subsection{The genetic approach}
 
-Work in progress.
+The genetic algorithm is a probabilistic global optimization metaheuristic. That means that it can be used to optimize solutions with many local optimums.
+It does this by simulating the genetic evolution of a population. The algorithm consists of a sequence of operations that can be run continuously until a solution
+presenting a sufficient accuracy has been created.
+
+
+\begin{enumerate}
+   \item The first step is to randomly initialize a population of solutions
+   \item We then evaluate all of the solutions present in the population and the algorithm returns a solution if its fitness value is small enough
+   \item Each set of parents for the next generation are chosen based off of a tourney approach (a fixed amount of solutions are randomly selected and the best two solutions are
+   chosen to be parents)
+   \item Each parent is then encoded into a table (Each parameter is condensed together to form a string)
+   \item Two children are then created by a uniform crossover method between both encoded parents
+   \item Each child then has a chance to have one value of its table mutated
+   \item The two children are then decoded and added to the next generation
+   \item Steps 3-7 are repeated until the next generation is the same size as the original generation
+   \item The original generation is replaced by the new generation
+   \item Steps 2-9 are repeated until a certain amount of generations has been created or a valid solution has been found.
+   \item The best solution is returned
+\end{enumerate}
+
+There are many methods for selecting the parents. One of the most popular methods is the roulette wheel solution where a solutions chance of being chosen is decided by its fitness
+divided by the sum of the fitnesses of the whole population. Another popular approach is the tournament selection where we randomly select a certain amount of solutions and pick the two
+with the best fitness. There are also many crossover approaches. The one we have chosen is a uniform crossover that means that each character of the encoded parents has an equal chance
+to be present in a child. The simplest approach is a one-point crossover in which an index is chosen for both parents and both encoded arrays representing the parents
+are swapped around those points to create the children.
+
+
+Genetic algorithms present a few problems such as premature convergence. This means that the algorithm can sometimes converge towards a point which is not the global
+optimum. We have implemented a variant of the genetic algorithm called "elitist selection" in order to reduce the likelihood of this happening. This means that the best solutions from a population are directly transferred
```

Comment:
Premature convergence is when the algorithm converges on an inaccurate solution. Elitist selection increases the accuracy of the algorithm and I haven't had this problem since.

Reply:
That's my point, premature convergence means that we converge too fast on a local optimum. If we use elite solutions we will start converging towards their solution which is potentially a local optimum. 

I would agree with you about the random crossover though, using a uniform crossover helps with the diversity of solutions rather than the accuracy. 

I found this link online that explains my point : https://ai.stackexchange.com/questions/3956/is-elitism-preferred-over-non-elitism-in-the-cross-over-operator


> The problem with pure elitism is that it causes the genetic algorithm to converge to local maxima instead of the global maxima. Basically pure elitism is just a race to the nearest local maxima and and as soon as you get to the local maxima, if you continue with elitism, you get almost no improved from there.

Reply:
I agree with you when it comes to pure elitism, which would essentially turn the genetic algorithm into a hill climbing algorithm , but in this case I am only using an elite group of two members which simply helps the algorithm converge into the right direction

Reply:
I'm sorry, the attached link did not really convey my point. What I meant is that relying solely on elitism does not help escaping local optimum. Rather, diversification is necessary. But on the other side, excessive diversification can impede convergence.

The main issue in genetic algorithm is achieving the proper balance between diversification and intensification. If we have too much diversification, then the algorithm will struggle to find a satisfactory solution. On the other hand, if intensification is overemphasized, the algorithm will converge toward the first local optimum it sees. 

Adding a pool of elite solutions looks like a very good idea, but from my point of view, it enhances the intensification aspect. By retaining good solutions, we can explore local optimum more thoroughly and remember them. 

I don't think elitism helps evade local optimum since this aspect of the algorithm does not promote diversification which is typically needed to avoid them and identify new optimum.

I'm not sure if I make sense. We can discuss this further on Slack if you want to.

Reply:
I agree that finding the right balance is optimal and that if the intensification is overemphasized it can cause the algorithm to get stuck in a local optimum . I believe the current elitist size is small enough that this isn't the case but it will be possible to verify this with the next test that is currently in development (Curve fitting) . I will benchmark the accuracy and performance to accurately judge the effect of our elitist selection. I look forward to continuing this discussion with you once i have the results.